python/lsst/cp/pipe/cpFunctions.py

```diff
@@ -0,0 +1,682 @@
+# This file is part of cp_pipe.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import datetime
+import numpy as np
+import time
+
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+import lsst.pipe.base.connectionTypes as cT
+import lsst.afw.math as afwMath
+import lsst.afw.image as afwImage
+from astro_metadata_translator import merge_headers, ObservationGroup
+from astro_metadata_translator.serialize import dates_to_fits
+from lsst.daf.butler import DatasetType
+
+
+class BlessCalibration(pipeBase.Task):
+    """Create a way to bless existing calibration products.
+
+    The inputs are assumed to have been constructed via cp_pipe, and
+    already exist in the butler.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        Butler repository to use.
+    inputCollection : `str`
+        Data collection to pull calibrations from.
+    outputCollection : `str`
+        Data collection to store final calibrations.
+    """
+    _DefaultName = 'BlessCalibration'
+    ConfigClass = pexConfig.Config
+
+    def __init__(self, *, butler=None, inputCollection=None, outputCollection=None,
+                 **kwargs):
+        super().__init__(**kwargs)
+        if butler is not None:
```

Comment:
Can this `Task` function at all if `butler is None`? If you're allowing for that possibility, you need to trap this case in the methods that rely on `self.butler` and `self.registry`.

Similarly for `inputCollection` and `outputCollection`.